### ISSUE TITLE
Confusion about GP.predict behavior and usage of `likelihood`

### ISSUE THREAD

**Issue by clairecw** [SheffieldML/GPy#886]
Hi!
Bit of a GP newbie here, so I apologize in advance for trivial questions.
I'm trying to use GPy as a classifier of the form (mu, RBF(ARD=True)), where mu is also a learned parameter. I'm using EP as the inference method. My data consists of 13 features with strictly binary (0/1) labels.

My first question is: after a couple .optimize() steps, it seems the learned mu is negative, despite no training labels falling outside [0, 1]. I figured this might be because my data is imbalanced in favor of negative labels, but even after upsampling, this is still the case. This result seems quite counterintuitive to me, but Is this normal/expected? 

I then figure this is where the likelihood function comes into play, as it makes use of the predicted variance values (and it appears that calls to `predict` produce non-negative values). For my use case, I'm not quite sure about the following:
1. Whether or not to use the likelihood function when calling `GP.predict`.
2. If yes to the above, which likelihood of those implemented to use. (I see that the default is Bernoulli, but I'm not sure how to determine whether that's appropriate for my data or not.)

Let me know if I should provide further details. Any help or pointers to resources would be much appreciated! 
Thanks.

**Comment by lawrennd** [comment]
Hi Claire,

Great question.

I'm presuming mu is a mean function here?

Great questions. I've also noticed this tendency with classification. So it's not unusual. I don't remember whether I ever came up with a good explanation ... 

Do you have a minimal example of your code to share as a gist?

I'm curious about a few things, for example whether you're using a bias kernel. 

The Bernoulli likelihood should be appropriate (unless there's something unusual about the data).

Neil

**Comment by clairecw** [comment]
Hey Neil,
Thanks so much for the prompt follow-up! Yes, mu is a mean function, we're currently using a constant mapping (so a singular scalar parameter).
Here is a minimal example, specifying the exact mean and kernel we're using: 
```
k = GPy.kern.RBF(len(features), ARD=True)
mf = GPy.mappings.constant.Constant(len(features), 1)
lik = GPy.likelihoods.Bernoulli()
m = GPy.core.GP(X=x_train
                Y=y_train, 
                kernel=k, 
                inference_method=GPy.inference.latent_function_inference.expectation_propagation.EP(),
                likelihood=lik,
                mean_function=mf
               )
```
Let me know if I should provide further details. 

I don't _think_ there's something unusual about our data, but would you happen to have some examples of oddities that might make Bernoulli an inappropriate choice?

**Comment by lawrennd** [comment]
Great, could I also ask (in the meantime) that you try with 

k = = GPy.kern.RBF(len(features), ARD=True) + GPy.kern.Bias(len(features))

And without the constant mean function? That's like a Bayesian equivalent of the model above.

If you do that, what is the Bias kernel's variance parameter?

You could also try with a constant mean function and the bias kernel ... that has an interpretation as an offset Gaussian prior over the mean function ... I'm also sort of curious what would happen with that!

**Comment by clairecw** [comment]
Thanks for the suggestions! 
Oh interesting - I get a bias variance of 11.85 - am I correct in assuming that this takes the place of the variable mean in my original model? If so, is it unexpected that this value also lies outside the [0, 1] range seen in the training data's labels (at least, it seems counterintuitive to me)? 

I'm now trying with both the constant mean with bias kernel - will update with results after :)

**Comment by lawrennd** [comment]
So one way to think about both these parameters is as parameters of the likelihood ... well the 'link function' at any rate. 

So in effect the model is changing the probit link to have an offset and uncertainty on that offset. Not sure if that helps with interpretation!

**Comment by clairecw** [comment]
Thanks so much Neil, that does help - just to ensure I have the right understanding, is it that the GP's output probits follow a 0-mean distribution, and the Bias kernel's variance parameter represents the aforementioned offset? How do I know what the offset uncertainty is? 

with `k = GPy.kern.RBF(len(features), ARD=True) + GPy.kern.Bias(len(features))`, I'm seeing once again a negative mean function constant, and bias variance close to 0.

**Comment by lawrennd** [comment]
So the link function is probit, which means the inverse link is the cumulative Gaussian (we'll denote that g).

This means probability of positive class, p, as a function of inputs, x, (which we'll denote p(x)) is given by

p(x) = g(f(x) + b)

Where f(x) is the output of the Gaussian process plus a mean function, b, which here is not a function of x because it's constant.

Adding a bias kernel is equivalent to 'gaussian uncertainty on b'. So if you use your model where your mean function is equal to mu, and your bias variance is equal to \alpha_b then you have

p(x) = g(f(x) + b)
b \sim N(mu, \alpha_b)

as your prior model. 

Does that make sense?

**Comment by clairecw** [comment]
Sorry, I think I'm only understanding partially. By "link function" are you referring to the likelihood function used?
And why is it that p(x) is a result of the _inverse_ link, rather than the link function itself?

I believe I do understand the uncertainty part now. So then the higher the bias kernel's variance parameter, the higher the uncertainty? And this uncertainty is only factored in when `include_likelihood=True`?

**Comment by lawrennd** [comment]
Questions answered in reverse order!

3) Well, no because it's not being included in the likelihood here, but in the kernel, it will be seen even without the likelihood. 

2) Yes, higher the bias kernel's variance parameter, the higher the uncertainty.

1) The link function is defined by statisticians. It transforms the parameter of interest (such as a positive rate in a Poisson mode, or a probability in a binomial likelihood) to from the parameters space (e.g.$\in [0, 1]$ for probability) to the full real line. 

This means you can write a linear model in the following form, e.g. for the log link function and $\lambda(t)$ as a rate changing over time,

$\log \lambda(t) = f(t) + g(t) $

Here saying the log of the rate is a linear function which is the sum of two other functions, $f(t)$ and $g(t)$.

In ML, it's more normal to think about the 'inverse link function', which in the log case is the exponential, and to write the rate in the following way.

$\lambda(t) = \exp(f(t) + g(t))$

For the probit model the link function is the inverse cumulative Gaussian. I found a PDF on line that roughly describes all this here: http://web.pdx.edu/~newsomj/mvclass/ho_link.pdf

But note I find that ML people are wont to use the term quite sloppily and refer to the $\exp()$ as the link function, statisticians call $\log()$ the link function for that case. I try to be precise to avoid confusion with the statisticians. Sometimes I call $\exp()$ the transformation function ... but I don't think that's standard.

**Comment by clairecw** [comment]
Thanks for the information! So returning to my initial questions, 

1. Is the model's uncertainty the reason I'm observing these negative mean functions or biases? (again, for context, none of my training labels fall outside the range [0, 1], so it doesn't make sense to me that the model's mean after fitting to this data should be outside this range)
2. What's the purpose of the likelihood function? (and related, when should I opt out of using it?)

**Comment by lawrennd** [comment]
We're getting to the point where I have to speculate about your data to know the right answer.

The likelihood function is to map the Gaussian process from continuous to the discrete values. So yes, you need that.

**Comment by clairecw** [comment]
Gotcha. Thanks so much Neil, I appreciate all the support.
I'll do some research and look into the source code myself, but in the off chance you have an answer off the top of your head - do you know, generally, how to best determine which likelihood function (out of Bernoulli, Gaussian, etc.) to use?

**Comment by lawrennd** [comment]
It depends on the noise you expect on data. 

Gaussian is standard for regression ... but data with out liers you use something heavy tailed like Student-t. 

For regressing the median you can use a Laplacian (or double exponential distribution).

Bernoulli is standard for classification. But like any model where the parameters are not on the real line, you need to also choose the link function. probit is standard for GPs because it makes the updates more analytically tractable. In general though you'll see people doing logit ... which leads to logistic regression.

For count data you often use Poisson models, or negative binomial if you think there's some 'over dispersion'. These models use link functions and a log link funciton is often used (implying a multiplicative model ...)

It's quite a big topic. Looking up "Generalised Linear Models" will give you some of the references.

**Comment by clairecw** [comment]
Thank you very much! This was very helpful and informative. Appreciate the time!
